pkg/preflight/checks.go

```diff
@@ -62,51 +62,139 @@ func (c CPUCheck) Run() (msg string, err error) {
 	return
 }
 
-func (c MemoryCheck) Run() (msg string, err error) {
-	meminfo, err := os.Open(procMemInfo)
-	if err != nil {
-		return
-	}
-	defer meminfo.Close()
-	scanner := bufio.NewScanner(meminfo)
-	var memTotalKiB int
-	for scanner.Scan() {
-		if n, _ := fmt.Sscanf(scanner.Text(), "MemTotal: %d kB", &memTotalKiB); n == 1 {
-			break
+func (c MemoryCheck) Run() (string, error) {
+	// We're working in KiB because that's what the fallback /proc/meminfo uses
+	var memTotalKiB uint
+	var wiggleRoom float32 = 1.0
+
+	// dmidecode is part of sle-micro-rancher, see e.g.
+	// https://build.opensuse.org/projects/SUSE:SLE-15-SP4:Update:Products:Micro54/packages/SLE-Micro-Rancher/files/SLE-Micro-Rancher.kiwi?expand=1
+	//
+	// The output of `dmidecode -t 19` will include one or more
+	// Memory Array Mapped Address blocks, for example on a system
+	// with 512GiB RAM, we might see this:
+	//
+	//	# dmidecode 3.5
+	//	Getting SMBIOS data from sysfs.
+	//	SMBIOS 2.8 present.
+	//
+	//	Handle 0x0024, DMI type 19, 31 bytes
+	//	Memory Array Mapped Address
+	//		Starting Address: 0x00000000000
+	//		Ending Address: 0x0007FFFFFFF
+	//		Range Size: 2 GB
+	//		Physical Array Handle: 0x000A
+	//		Partition Width: 1
+	//
+	//	Handle 0x0025, DMI type 19, 31 bytes
+	//	Memory Array Mapped Address
+	//		Starting Address: 0x0000000100000000k
+	//		Ending Address: 0x000000807FFFFFFFk
+	//		Range Size: 510 GB
+	//		Physical Array Handle: 0x000B
+	//		Partition Width: 1
+	//
+	// By adding together all the "Range Size" lines we can determine
+	// the amount of physical RAM installed.  Note that it's possible
+	// for units to be specified in any of "bytes", "kB", "MB", "GB",
+	// "TB", "PB", "EB", "ZB", so we have to handle all of them...
+	// (see http://git.savannah.nongnu.org/cgit/dmidecode.git/tree/dmidecode.c#n283)
+	out, err := execCommand("/usr/sbin/dmidecode", "-t", "19").Output()
+	if err == nil {
+		rangeSizeToKiB := func(rangeSize uint, unit string) uint {
+			switch unit {
+			case "GB":
+				// We're probably usually going to see GB
+				return rangeSize << 20
+			case "MB":
+				// This seems unlikely
+				return rangeSize << 10
+			case "kB":
+				// This seems even more unlikely
+				return rangeSize
+			case "bytes":
+				// Seriously, are you kidding me?
+				return rangeSize >> 10
+			}
+			return 0
+		}
+
+		for _, line := range strings.Split(string(out), "\n") {
+			var rangeSize uint
+			var unit string
+			if n, _ := fmt.Sscanf(strings.TrimSpace(line), "Range Size: %d %s", &rangeSize, &unit); n == 2 {
+				if unit == "TB" || unit == "PB" || unit == "EB" || unit == "ZB" {
+					// If we've somehow got a Memory Array Mapped Address
+					// with one of these enormous units, let's just pretend
+					// we've got a terabyte of RAM and be done with it ;-)
```

Comment:
Could we add some logs to record this situation?
I know that might not happen very soon, but we could add a log at this time.

Reply:
Done